### ISSUE TITLE
[Bug]: No page limit when sorting by TopAll

### ISSUE THREAD

**Issue by glibg10b** [LemmyNet/lemmy#3680]
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Summary

You can visit any page when sorting by TopAll, up to [the last page](https://lemmy.ml/?dataType=Post&listingType=All&page=15894&sort=TopAll) (as of the time of writing). Loading this page takes a while, which I imagine can be exploited to bring a server down.

Feel free to close this if this is a non-issue.

### Steps to Reproduce

1. Visit https://lemmy.ml/?dataType=Post&listingType=All&page=<page\>&sort=TopAll, where \<page\> is something big (above 15000)
2. Observe how long it takes to load

### Technical Details

After 15894 (as of the time of writing), it immediately errors out

### Version

BE 0.18.2

**Comment by googleben** [comment]
My understanding, and I'm far from a DB person so I may be wrong, is that instead of using `offset` as Lemmy currently does (`offset = (page-1)*limit`), you can increase performance by doing e.g. `WHERE some_sorting_metric < cursor` where `cursor` is the final value of the metric on the last page. So if you sort by top with a limit of 10, and the last result on page 1 has a vote total of 100, you do `WHERE score < 100` to start at the 11th result.

You would also need to do other things, like I think using composite types in the comparison with an ID cursor in them to handle things like posts with the same number of votes through a page boundary, at which point I'm well out of my depth with regards to the performance impact. So no clue how well that would work. 

If that does improve performance, though, then the API could only allow relatively small `offset` values and require cursor values for deep scrolling, so the best of both worlds. On the other hand this would require updating all Lemmy API clients so not preferable if there's a less impactful solution. 

For reference, the query in question is constructed [here](https://github.com/LemmyNet/lemmy/blob/cf2229d66520fdb3833c4f1b1d3aa569f41d8883/crates/db_views/src/post_view.rs#L219) (`lemmy_db_views::post_view::PostQuery::list`)) and the API endpoint is `/api/v3/post/list`. A limit on the page number could be implemented [here](https://github.com/LemmyNet/lemmy/blob/cf2229d66520fdb3833c4f1b1d3aa569f41d8883/crates/db_schema/src/utils.rs#L159) (`lemmy_db_schema::utils::limit_and_offset`) which would also impose the same limit for comments and (I think) the community list, solving the same potential problem in those places as well - though that's less important since there are way more posts than comments (on any given post) or communities.

**Comment by googleben** [comment]
Alright, I just did a very rudimentary test. I set up a test DB with 60,000 generated posts with random `published` fields any time in the last 28 days. Here are some queries with their rough timing information from my testing:

```SQL
SELECT * FROM post ORDER BY published, id LIMIT 20;
Time: 17.5 ms +/- 0.5 ms
SELECT * FROM post ORDER BY published, id LIMIT 20 OFFSET 20*999;
Time: 35 ms +/- 5 ms
SELECT * FROM post WHERE published > '2023-07-07 08:56:02.725363' ORDER BY published, id LIMIT 20;
Time: 14.5 ms +/- 1 ms
SELECT * FROM post WHERE (published, id) > ('2023-07-07 08:56:02.725363', 2619) ORDER BY published, id LIMIT 20;
Time: 14.5 ms +/- 1 ms
```

For some more info, here's the output of `EXPLAIN` for queries 1, 2, and 3, supporting the timing results:

```SQL
EXPLAIN SELECT * FROM post ORDER BY published, id LIMIT 20 OFFSET 999*20;
-[ RECORD 1 ]--------------------------------------------------------------------------------------
QUERY PLAN | Limit  (cost=19576.61..19578.91 rows=20 width=780)
-[ RECORD 2 ]--------------------------------------------------------------------------------------
QUERY PLAN |   ->  Gather Merge  (cost=17278.91..21337.72 rows=35294 width=780)
-[ RECORD 3 ]--------------------------------------------------------------------------------------
QUERY PLAN |         Workers Planned: 1
-[ RECORD 4 ]--------------------------------------------------------------------------------------
QUERY PLAN |         ->  Sort  (cost=16278.90..16367.13 rows=35294 width=780)
-[ RECORD 5 ]--------------------------------------------------------------------------------------
QUERY PLAN |               Sort Key: published, id
-[ RECORD 6 ]--------------------------------------------------------------------------------------
QUERY PLAN |               ->  Parallel Seq Scan on post  (cost=0.00..1425.94 rows=35294 width=780)

EXPLAIN SELECT * FROM post ORDER BY published, id LIMIT 20;
-[ RECORD 1 ]-----------------------------------------------------------------------
QUERY PLAN | Limit  (cost=3269.58..3269.63 rows=20 width=780)
-[ RECORD 2 ]-----------------------------------------------------------------------
QUERY PLAN |   ->  Sort  (cost=3269.58..3419.58 rows=60000 width=780)
-[ RECORD 3 ]-----------------------------------------------------------------------
QUERY PLAN |         Sort Key: published, id
-[ RECORD 4 ]-----------------------------------------------------------------------
QUERY PLAN |         ->  Seq Scan on post  (cost=0.00..1673.00 rows=60000 width=780)

EXPLAIN SELECT * FROM post WHERE (published, id) > ('2023-07-07 08:56:02.725363', 2619) ORDER BY published, id LIMIT 20;
-[ RECORD 1 ]-----------------------------------------------------------------------------------------------------------------
QUERY PLAN | Limit  (cost=3031.08..3031.13 rows=20 width=780)
-[ RECORD 2 ]-----------------------------------------------------------------------------------------------------------------
QUERY PLAN |   ->  Sort  (cost=3031.08..3130.49 rows=39763 width=780)
-[ RECORD 3 ]-----------------------------------------------------------------------------------------------------------------
QUERY PLAN |         Sort Key: published, id
-[ RECORD 4 ]-----------------------------------------------------------------------------------------------------------------
QUERY PLAN |         ->  Seq Scan on post  (cost=0.00..1973.00 rows=39763 width=780)
-[ RECORD 5 ]-----------------------------------------------------------------------------------------------------------------
QUERY PLAN |               Filter: (ROW(published, id) > ROW('2023-07-07 08:56:02.725363'::timestamp without time zone, 2619))
```

As expected, large `offset` values kill performance and using cursors instead seems to keep performance roughly on par with a "first page" query. Looks like it ought to be a viable solution as long as the rest of the actual query doesn't mess with things, which I don't think it should. The only problem is that it messes with the API, as far as I can see.

A couple of notes for completeness: 

- The publish time and post id were the values of the first post in the `20*999` query, so roughly the same results, just offset by a single post which shouldn't matter.  
- Page 999 of 20 results each happened to be at row ~19980 from the start of the sorted table, almost exactly 1/3 of the way through the table, so this test should be pretty representative of deep scrolling.
- I don't think it should matter besides inflating times, but my DB is stored on a hard drive since I don't have a ton of SSD space left.

**Comment by Nutomic** [comment]
Should be fixed by cursor-based pagination in https://github.com/LemmyNet/lemmy/pull/3872.
